// pesepay/dist/payments/payment.js
"use strict";
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Payment = void 0;
var Payment = (function () {
    function Payment(currencyCode, paymentMethodCode, customer) {
        this.currencyCode = currencyCode;
        this.paymentMethodCode = paymentMethodCode;
        this.customer = customer;
    }
    Payment.prototype.setRequiredFields = function (_a) {
        var requiredFields = __rest(_a, []);
        this.paymentMethodRequiredFields = requiredFields;
        this.paymentRequestFields = requiredFields;
    };
    return Payment;
}());
exports.Payment = Payment;
